File: src/main/java/dev/psygamer/wireframe/util/collection/FreezableTreeMap.java
package dev.psygamer.wireframe.util.collection;

import java.util.*;

public class FreezableTreeMap <K, V> extends FreezableMap<K, V> {
	
	public FreezableTreeMap() {
		super(new TreeMap<>());
	}
	
	public FreezableTreeMap(final Map<? extends K, ? extends V> map) {
		super(new TreeMap<>(map));
	}
	
	public FreezableTreeMap(final SortedMap<? extends K, ? extends V> sortedMap) {
		super(new TreeMap<>((SortedMap<K, ? extends V>) sortedMap));
	}
	
	public FreezableTreeMap(final Comparator<? super K> comparator) {
		super(new TreeMap<>(comparator));
	}
	
	@Override
	public FreezableMap<K, V> copy() {
		final FreezableMap<K, V> map = new FreezableTreeMap<>(this);
		
		if (isFrozen())
			map.freeze();
		
		return map;
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/util/collection/FreezableTreeMap.kt
package dev.psygamer.wireframe.util.collection

import java.util.*

class FreezableTreeMap<K, V> : FreezableMap<K, V> {

	constructor() : super(TreeMap<K, V>())
	constructor(map: Map<out K, V>) : super(TreeMap<K, V>(map))
	constructor(sortedMap: SortedMap<out K, out V>) : super(TreeMap<K, V>(sortedMap))
	constructor(comparator: Comparator<in K>) : super(TreeMap<K, V>(comparator))
}